### ISSUE TITLE
Switch EpochSlots to be frozen slots, not completed slots

### ISSUE THREAD

**Issue by t-neIson** [solana-labs-community/solana#149]
#### Problem
If the cluster doesn't directly voted on one version of a slot `S` (maybe only voted on the descendants), and we marked slot `S` as dead, then currently there's no mechanism by which to detect that the cluster has frozen another version of this slot.

#### Summary of Changes
Switch EpochSlots to be for frozen slots instead of just completed slots, so we can later invoke dump/repair logic if a slot is frozen > duplicate threshold of the cluster. Design follows the proposal here: f79c0e8#diff-0bd0e50030bdee14082c678da5f3098df8156825fa86c59ad55ddbd2a3669585R82, which describes the context in more detail

Fixes #

**Comment by sakeridge** [comment]
Ok, so I guess the only downside might be that a node which is trying to repair a slot which is errored wouldn't have as much information to repair since it doesn't know who has the slot data.

I guess we don't need any kind of switch-over. The EpochSlot still has the same data and meaning for being able to repair a successfully frozen slot.

**Comment by mvilnes** [comment]
Yeah people repairing an errored slot will not have all the details, and now people who want to repair non-errored slots will have to wait for people to replay everything which might cause a little more of a delay as well, usually 200-300ms (difference between slot complete and bank frozen times)

**Comment by CrieasofCarrots** [comment]
Look good to me. Did you consider having the leader push their slots into epoch slots too? If they are the one producing the dupe blocks it's a little weird but it would help cross the threshold. Seems useful for detecting descendant confirmed blocks too

**Comment by jeffweshington** [comment]
This should run on the leader as well, when they freeze the bank, they'll also update their own EpochSlots

**Comment by t-neIson** [comment]
Ah, that's right. I forgot that replay stage is what advances the working bank in poh service for leader blocks, not banking stage itself.
